kbs/docs/initdata.md

```diff
@@ -0,0 +1,267 @@
+of TEE hardware) or software vendor (like guest kernel for VMs). These
+components are relatively static, which means they may be the same among
+multiple deployments
+
+In some scenarios, users would inject some other information like
+[policy files for kata](https://github.com/kata-containers/kata-containers/blob/main/docs/how-to/how-to-use-the-kata-agent-policy.md),
+[configuration files for components running in guest](https://github.com/confidential-containers/guest-components/tree/main/confidential-data-hub#configuration-file),
+[identity files to specify the identity of TEE](https://github.com/keylime/rust-keylime/blob/master/keylime-agent.conf)
+into the TEE guest when launching.
+
+Compared with static software running in TEE (like guest firmware for TDX VM,
+libos for SGX enclave), these information changes dynamically between different
+deployments and are usually configurations
+We call these information or configurations _Initdata_. Initdata mechanism will
+provides a way to protect their integrity by remote attestation. One thing to note
+is that the confidentiality will not be protected by initdata mechanism because
+the untrusted host can see the plaintext of the data.
+
+To achieve this goal, we defined the following things
+- A data structure named **Initdata Metadata**. This structure is provided by the
+user to contain any initdata in key-value format to untrusted host to inject into
+the TEE when launching. We do not limit the encoding of this data structure, which
+means that JSON and TOML are optional. This will be introduced in [Initdata Metadata](#initdata-metadata)
+- A data integrity binding mechanism. It will guide the untrusted host to bind the
+digest of the `data`` part in the Initdata Metadata to the hardware TEE-specific 
+field in evidence. This field will be checked by the verifier during the remote
+attestation. This will be introduced in 
+[Integrity Binding for Different TEEs](#integrity-binding-for-different-tees)
+- A data serialization method. This method is used to serialize the Initdata Metadata
+into a canonicalized one, which will help to get consistent cryptographic hash for
+different forms of a same Initdata Metadata. This will be introduced in
+[Data Canonicalization Algorithm](#data-canonicalization-algorithm).
+
+This spec does not define how the Initdata Metadata will be delivered into the TEE.
+Different projects will have its own way to do this. For Confidential Containers,
+we will use kata-runtime and kata-agent to collaborate to achieve this function.
+
+## Terminology
+
+This section will introduce the terminology used in this spec to avoid ambiguity.
+
+- `Initdata`: Data that needs to be injected when the TEE is started. This data
+requires integrity protection but does not require confidentiality protection.
+- `Initdata Metadata`: A data structure that includes initdata and other information
+that will help to calculate the initdata digest.
```

Comment:
Sounds good. The term definition also annoys me.

Some confusing items:
- initdata: data to be injected
- initdata metadata: `version`, `algorithm` and probably `digest` (if we decide to use it)
- structured init data: initdata + metadata
- initdata field/hostdata: the one inside evidence
- initdata digest: digest of initdata calculated following the spec

Does these make sense?

Reply:
I am fine with that terminology. There is also an alternative

* initdata: the entire structure (i.e. structured init data)
* initdata metadata: just the metadata fields (same as above)
* initdata data/fields: everything in the data section
* hw initdata / hostdata: the field in the hw evidence
* initdata digest: same

I don't have any preference. I guess the first term and the third one are kind of interchangeable.

Reply:
Ok. To align with the API of `CheckInitdata`  in AA and `Initdata` in AS, I prefer hw initdata to refer to the field inside HW evidence

Reply:
Changed to
- `Initdata`: A data structure that includes initdata data and other information
that will help to calculate the initdata digest. The whole data structure will
be delivered into the guest.
- `Initdata Metadata`: Metadata fields of initdata, s.t. `algorithm`, `version`,
etc. They are used to calculate the initdata digest.
- `Initdata data`: Data that needs to be injected when the TEE is started. This data
requires integrity protection but does not require confidentiality protection. In
initdata, this will be included inside the `data` section.
- `Initdata digest`: Digest of the initdata data calculated following this spec.
It will be used as the value of the TEE hostdata/initdata field.
- `TEE initdata/hostdata`: Fields that can be bound to a specific TEE instance. This field
information will be included in the TEE-signed remote attestation report. Typically,
Intel TDX's `mr_config_id`, AMD SNP's `hostdata` and Arm CCA's `CCA_REALM_PERSONALIZATION_VALUE`.
In order to avoid confusion with the hostdata field of AMD SNP, when we do not
emphasize a specific platform s.t. SNP, we are referring to the corresponding fields of
various TEE platforms.